src/ipc/ipc4/helper.c

```diff
@@ -371,28 +371,116 @@ int ipc_buffer_free(struct ipc *ipc, uint32_t buffer_id)
 int ipc_comp_connect(struct ipc *ipc, ipc_pipe_comp_connect *_connect)
 {
 	int dir = PPL_CONN_DIR_BUFFER_TO_COMP;
-	struct comp_dev *src, *dst;
-	struct comp_buffer *buffer;
+	struct ipc4_module_bind_unbind *bu;
+	struct comp_buffer *buffer = NULL;
+	struct comp_dev *src, *sink;
+	struct list_item *sink_list;
+	int src_id, sink_id;
 	uint32_t flags;
+	int ret;
+
+	bu = (struct ipc4_module_bind_unbind *)_connect;
+	src_id = IPC4_COMP_ID(bu->header.r.module_id, bu->header.r.instance_id);
+	sink_id = IPC4_COMP_ID(bu->data.r.dst_module_id, bu->data.r.dst_instance_id);
+	src = ipc4_get_comp_dev(src_id);
+	sink = ipc4_get_comp_dev(sink_id);
+	if (!src || !sink) {
+		tr_err(&ipc_tr, "failed to find src %x, or dst %x", src_id, sink_id);
+		return IPC4_INVALID_RESOURCE_ID;
+	}
 
-	src = ipc4_get_comp_dev(_connect[0]);
-	dst = ipc4_get_comp_dev(_connect[1]);
+	list_for_item(sink_list, &src->bsink_list) {
+		int src_queue = bu->data.r.src_queue;
+		struct comp_buffer *buf;
 
-	if (!src || !dst) {
-		tr_err(&ipc_tr, "failed to find src %x, or dst %x", _connect[0], _connect[1]);
+		buf = container_of(sink_list, struct comp_buffer, source_list);
+		if (buf->id == src_queue) {
+			buffer = buf;
+			break;
+		}
+	}
+
+	if (!buffer)
 		return IPC4_INVALID_RESOURCE_ID;
+
+	irq_local_disable(flags);
+	list_item_prepend(buffer_comp_list(buffer, dir), comp_buffer_list(sink, dir));
+	buffer_set_comp(buffer, sink, dir);
+	comp_writeback(sink);
+	irq_local_enable(flags);
+
+	dcache_writeback_invalidate_region(buffer, sizeof(*buffer));
```

Comment:
doesnt this need to be done with IRQs off ?

Reply:
According to ipc_comp_to_buffer_connect, we don't need to do it. One buffer should be connected to only one component, so we don't need to.

Reply:
but we are writing back the component above with IRQs off, it would seem logical to align the cache here for both comp and buffer with IRQs off